### ISSUE TITLE
Remove x-swagger-router-controller

### ISSUE THREAD

**Issue by whitlockjc** [apigee-127/swagger-tools#221]
While discussing #218, I realized that we build swagger-router wrong due to us building against the moving Swagger 2.0 specification.  Basically, `operationId` is to be unique across all operations and with that, we have no ambiguity so that renders `x-swagger-router-controller` obsolete.  The idea of `x-swagger-router-controller` was to remove ambiguity so that you could have a `getById` in your `Project` controller and your `User` controller and wire them up properly.  Since that is not a valid use case, `x-swagger-router-controller` is pretty much useless now.

Since this changes the contract of things, we would need a patch release to support this but so does #203.

**Comment by earth2marsh** [comment]
:+1:

**Comment by theganyo** [comment]
Ack. If I'm understanding what you're proposing correctly, I believe it means that all controller function names are forced to be unique across the entire application.

**Comment by whitlockjc** [comment]
Well, the reason behind `x-swagger-router-controller` was to remove ambiguity and since technically we don't have any from a Swagger perspective, there's no need to have this in the Swagger document.  So in Swagger land, I cannot technically have multiple operations with the same `operationId`.

From a JavaScript perspective, things are a little different.  If you use the example above, you would have two potential implementations for the `getById` `operationId`.  That's the first issue.  The second is that I do not think that Swagger should have any impact on my controllers themselves because I may in fact want to have `Project#getById` and `User#getById`.  In cases like this, you'd have to build your `controllers` map yourself, much like you have to do now when you want to not use the convention.

So this change would basically change the convention but would still use the same, manual approach for working around the convention.  Make sense?

**Comment by theganyo** [comment]
Sure. I'm just saying that this is a pretty serious breaking change for existing users.

**Comment by whitlockjc** [comment]
No doubt, I definitely don't take this lightly.  A change like this, and #203, change the current interface/conventions and we'd need to do a patch release.  _(We should probably cut a 1.0 release...)_

**Comment by theganyo** [comment]
Technically, I believe this should be a "major" release according to semantic versioning as it breaks existing behavior.

**Comment by bachp** [comment]
I agree that this is a breaking change and we should probably keep the old behaviour as deprecated until the 1.0 release. Kind of how it is done in #218.

BTW: The breaking change rule does not apply to 0.x releases, from the [semver](http://semver.org/spec/v2.0.0.html) spec:

> Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.

**Comment by theganyo** [comment]
True, it's all "wild west" in the 0.x... but that won't make users feel any better when you break them. :)

In any case, the main problem isn't adding the new behavior... we can do that and deprecate the old without breaking anything. It's the validation. We need to validate that operationId is unique across the entire spec - and we must assume that it isn't for current users.

**Comment by jcferrer** [comment]
It will be a breaking change for my team for sure. We already have multiple swagger definition files which rely in the x-swagger-router-controller and operationId combination. How are the swagger tools going to handle the dispatching if the controller extension is removed?. When you say the operationId is to be unique, is there a naming convention for the operationId to now reference both controller and function to be called?.

I would like to know this upfront if possible so we can prepare for the change.

Thanks

**Comment by whitlockjc** [comment]
While discussing #218, I realized that we were not validating `operationId` per the spec.  As mentioned in #219, we need to validate that the `operationId` is unique across all operations.  So if the `operationId` is unique, it doesn't matter where in the controllers it's defined as long as the name matches.  Now of course, there can be some ambiguity.  _(See the example mentioned in #218)_  I think in that case we'd be better off printing warnings to the effect that there are multiple possible options for the route handler as identified by the `operationId`.

If I were to work on this now, and I'm not so you've got time of course, I would probably blow away the annotation as we know it.  I would continue to pass the `controllers` option to the swagger-router middleware but instead of using `{ControllerFileBase}_{exportedFunctionName}` as the wiring key, I'd just use the `{expoertedFunctionName}` and the controller from which it came no longer plays a role.

Since this is a breaking change, I'd rather discuss this at whatever length necessary to do this right.  In the end, I think when I work on #219, it will break a lot of deployments without mucking with `x-swagger-router-controller` because I'm sure people are violating the unique `operationId` and letting `x-swagger-router-controller` handle the ambiguity.

Please continue to add your thoughts/opinions on the matter.  I will not work on this, or #219, until we have come to some sort of an agreement.

**Comment by bachp** [comment]
As far as I understand from the specs the `operationId` already has to be unique accross an API definition. I'm not sure what is included in an API definition ([issue](https://github.com/swagger-api/swagger-spec/issues/381) raised for spec). But lets assume it's all the operations defined in the same file.

This would mean you need one controller per file. Or one controller for whatever an "API definition" maps to once clarified.

**Comment by whitlockjc** [comment]
Well, that's not entirely true as we _(swagger-tools)_ dictate how we find route handlers and map them.  Right now an API is a single Swagger definition document and whatever number of JavaScript controllers you deem fit.  Some people work with one controller, others have one controller per business object.  _(Example: Projects, Users, Workspaces, ...)_  I would hope that whatever decision we make does not change this flexibility as I'm definitely not a fan of being told how to structure my code.  _(Now I realize that we will likely have to impose some rules to allow for `operationId`-based mapping but that doesn't mean we change how people structure their code.)_

As for ambiguity, in a single controller approach, of course you do not need to worry about ambiguity because you can only export one function for a given name.  But if you have N different controllers then we could run into the situation where you can have N different exported functions with the same name across the controller base and we'd have to address this.

So using `operationId` as the mapping without `x-swagger-router-controller` makes sense from a Swagger perspective, from the implementation/JavaScript perspective, this can be a limitation.  So I'm hoping we can continue this discussion to see what makes sense.  The only offer on the table right now is the one I proposed and it was really just a way to get the dialog going.  It is not set in stone.

**Comment by whitlockjc** [comment]
By the way, thanks everyone for chiming in.  The hope here is to get a solution we can all agree upon and go from there instead of just doing what is obvious and forcing it upon you.

**Comment by joostfarla** [comment]
What I'm wondering is whether the `operationId` has to be used for mapping operations to controllers and methods in the first place. I definitely agree with @whitlockjc on the argument that Swagger shouldn't dictate developers about how they structure their routing/controllers/methods and which naming conventions they should use. Another important argument is that the reader of the spec does not have anything to do with implementation details. This is something you want to keep private as a developer.

Since `x-*` elements (extensions) are [filtered out](https://github.com/theganyo/swagger-node-runner/blob/master/lib/connect_middleware.js#L34) by some libraries (like `swagger-node-runner`), this would make them well-suited for adding implementation details to the spec which are only exposed to the API developer. On the other hand you could state that `x-*` elements do not always have to be kept hidden from the outside, since you might want to us them for extending the spec for some public purpose.

A possible approach could be to introduce some kind of generic `x-swagger-router-id` element as a replacement for `operationId`. How the library maps this ID to the corresponding controller/method can be standardized conforming some naming conventions, but it has to be really flexible to match all edge cases. The old elements could still be used, but they can be deprecated.
